src/icp_v2.cc

```diff
@@ -669,39 +655,56 @@ icpHandleUdp(int sock, void *)
             /* or maybe an EHOSTUNREACH "No route to host" message */
             if (xerrno != ECONNREFUSED && xerrno != EHOSTUNREACH)
 #endif
-                debugs(50, DBG_IMPORTANT, "icpHandleUdp: FD " << sock << " recvfrom: " << xstrerr(xerrno));
-
+                throw TextException(xstrerr(xerrno), Here());
             break;
         }
 
         ++(*N);
         icpCount(buf, RECV, (size_t) len, 0);
         buf[len] = '\0';
-        debugs(12, 4, "icpHandleUdp: FD " << sock << ": received " <<
-               (unsigned long int)len << " bytes from " << from);
-
-#ifdef ICP_PACKET_DUMP
 
-        icpPktDump(buf);
-#endif
-
-        if ((size_t) len < sizeof(icp_common_t)) {
-            debugs(12, 4, "icpHandleUdp: Ignoring too-small UDP packet");
-            break;
-        }
-
-        icp_version = (int) buf[1]; /* cheat! */
+        // XXX: possible unaligned integer access
+        auto pkt = reinterpret_cast<icp_common_t *>(buf);
+        auto payloadLength = ntohs(pkt->length);
+
+        if (static_cast<size_t>(len) < sizeof(icp_common_t))
+            throw TextException("too-small UDP packet (ignoring)", Here());
+
+        if (payloadLength < len-sizeof(icp_common_t))
+            throw TextException("length value larger than data received (ignoring)", Here());
+
+        // TODO use BinaryTokenizer parse to validate fully
+        if (pkt->opcode < 0 || pkt->opcode >= ICP_END)
+            throw TextException(ToSBuf("unknown opcode ", pkt->opcode), Here());
+
+        struct in_addr a;
+        a.s_addr = pkt->shostid;
+        debugs(12, 2, "ICP Client remote=" << from << " FD " << sock);
+        debugs(12, 2, "ICP Client REQUEST:\n---------\n" <<
+               "opcode=(" << std::setw(3) << pkt->opcode << ")" << icp_opcode_str[pkt->opcode] <<
+               ", version=" << pkt->version <<
+               ", length=" << payloadLength <<
+               ", reqnum=" << ntohl(pkt->reqnum) <<
+               ", flags=" << std::hex << ntohl(pkt->flags) << std::dec <<
+               ", shostid=" << Ip::Address(a) <<
+               Debug::Extra <<
+               Raw(nullptr, (buf+sizeof(icp_common_t)), payloadLength).minLevel(DBG_DATA).hex() <<
```

Comment:
This output is displaying ICP protocol syntax received (if valid), not HTTP or whatever protocol meta got sent in the payload. In Squid cache.log payload data is at level-9 / DBG_DATA.

The KISS relevance is that the ICP payload is being added to the same debugs() statement as the ICP header display (at the level-9). The slightly more complex variant would have it as a second debugs() statement with a repeated set of prefix connection details to identify it (in case of overlapping worker outputs).

Also, FTR I thought it was just URLs too initially but when you requested hex I looked at the RFC closely to see if that was necessary (it doubles the log increase per packet). Some of the packets do have non-URL data and binary blobs.

From the security / "safety" aspect, the principle being followed is to reduce the attack surface. In this case the surface area is number of places any potential sensitive info in the URL is available to attackers (CWE "Information Leak"). It is not *necessary* to log the ICP payload to perform basic traffic flow accounting / tracing, so it should not be done at the log level-2 provided for that. The only reason it is logged at all is for advance troubleshooting of URl mangling and normalization between proxies using ICP.

Reply:
These arguments are pretty much invalidated by the fact that ICP "payload" is not really payload in any true sense of that word when interpreted in relevant/Squid context. It is just a variable-size portion of the header. HTTP has one mechanism to supply such variable-size header information, PROXY protocol has another, ICP has its own way. Yes, the word is spelled "payload" in the RFC, but that is just letters. Sometimes, letters mislead if interpreted too literally. We need to think about the _meaning_ of the ICAP message part we are printing. And that meaning is no different than a part of an HTTP or PROXY protocol header. We should use the same approach to report all those parts.

The fact that some headers carry binary data is not relevant to this change request. That fact only affects how we render the data, and we are in agreement that it should be rendered using hex (unless somebody wants to invest even more time to make that decision based on message type and payload field contents).

But I give up. This problem is just not worth my time. If I failed to convince you, leave minLevel(). No need to explain why I have not convinced you.

----

> with a repeated set of prefix connection details to identify it (in case of overlapping worker outputs).

This is out of topic here, but just in case this resurfaces elsewhere: We should not repeat details. Squid cache.log format already identifies which worker the record is coming from.

Reply:
> These arguments are pretty much invalidated by the fact that ICP "payload" is not really payload in any true sense of that word when interpreted in relevant/Squid context. It is just a variable-size portion of the header.

The relevant context is that we are displaying a single message from a message-based protocol. Using that protocols terminology, in ICP specific code.

RFC 2816 section 1 is very explicit how to interpret the octets received:
"
ICP Message Format

   The ICP message format consists of a 20-octet fixed header plus a
   variable sized payload (see Figure 1).
"

The "Squid way" (er, context?) to display messages is level-2 for header, level-9 for payload.